Cabal/Distribution/Simple/GHC/Internal.hs

```diff
@@ -396,15 +434,50 @@ componentGhcOptions verbosity implInfo lbi bi clbi odir =
       ghcOptExtensionMap    = Map.fromList . compilerExtensions $ (compiler lbi)
     }
   where
-    toGhcOptimisation NoOptimisation      = mempty --TODO perhaps override?
-    toGhcOptimisation NormalOptimisation  = toFlag GhcNormalOptimisation
-    toGhcOptimisation MaximumOptimisation = toFlag GhcMaximumOptimisation
-
     exe_paths = [ componentBuildDir lbi (targetCLBI exe_tgt)
                 | uid <- componentExeDeps clbi
                 -- TODO: Ugh, localPkgDescr
                 , Just exe_tgt <- [unitIdTarget' (localPkgDescr lbi) lbi uid] ]
 
+toGhcOptimisation :: OptimisationLevel -> Flag GhcOptimisation
+toGhcOptimisation NoOptimisation      = mempty --TODO perhaps override?
+toGhcOptimisation NormalOptimisation  = toFlag GhcNormalOptimisation
+toGhcOptimisation MaximumOptimisation = toFlag GhcMaximumOptimisation
+
+
+componentCmmGhcOptions :: Verbosity -> GhcImplInfo -> LocalBuildInfo
+                      -> BuildInfo -> ComponentLocalBuildInfo
+                      -> FilePath -> FilePath
+                      -> GhcOptions
+componentCmmGhcOptions verbosity _implInfo lbi bi clbi odir filename =
+    mempty {
+      -- Respect -v0, but don't crank up verbosity on GHC if
+      -- Cabal verbosity is requested. For that, use --ghc-option=-v instead!
+      ghcOptVerbosity      = toFlag (min verbosity normal),
+      ghcOptMode           = toFlag GhcModeCompile,
+      ghcOptInputFiles     = toNubListR [filename],
+
+      ghcOptCppIncludePath = toNubListR $ [autogenComponentModulesDir lbi clbi
+                                          ,autogenPackageModulesDir lbi
+                                          ,odir]
+                                          -- includes relative to the package
+                                          ++ PD.includeDirs bi
+                                          -- potential includes generated by `configure'
+                                          -- in the build directory
+                                          ++ [buildDir lbi </> dir | dir <- PD.includeDirs bi],
+      ghcOptCppOptions     = cppOptions bi,
+      ghcOptCppIncludes    = toNubListR $
+                             [autogenComponentModulesDir lbi clbi </> cppHeaderName],
+      ghcOptHideAllPackages= toFlag True,
+      ghcOptPackageDBs     = withPackageDB lbi,
+      ghcOptPackages       = toNubListR $ mkGhcOptPackages clbi,
+      ghcOptOptimisation   = toGhcOptimisation (withOptimization lbi),
+      ghcOptDebugInfo      = toFlag (withDebugInfo lbi),
+      ghcOptExtra          = hcOptions GHC bi,
```

Comment:
Oops I forgot to fix this while copy-pasting from the other cases: s/hcOptions GHC bi/cmmOptions bi/
I guess it should fix cmm-options passing.

Reply:
yeah... this is the kind of bugs I was worried about... :-)

I've rebased and updated the PR to include your fixup suggestion; I'm currently testdriving this locally

I'm also going to un-DRAFT this PR to invite more code-review